skpr/mtk
# Add optional S3 export/import capability

## Overview

Currently MTK sends everything to stdout. In keeping in-line with the design pattern, this change provides an avenue to export data directly to S3. This is facilitated through the use of permissions linking RDS and S3 as described in the links at the bottom of this PR.

## Implementation

This PR adds two optional parameters which will manipulate the output and query when provided.

* `--data-export` is a boolean value which simply tells mtk the query created should be executing a `SELECT INTO OUTFILE` query.
* `--data-path` is a string providing the path to the data. This will output data to the configured directory by the mysql instance, so users can use MySQL queries to either write to the local file system or directly to S3 as described in the links at the bottom of this PR.
* `--region` is the bucket region, specified when importing data from the bucket.
* Existing logic is unchanged and unless both fields are configured, meaning existing users would not be impacted.

## How to Test

Currently the way to test this is to ensure the MySQL queries validate correctly, there is a configuration component that still needs to be explored.

* Start a docker container that you are able to connect to which features a populated database, this could be an application, project or just a vanilla mysql container.
* Build a fresh copy of MTK with `make build`
* Run the dump command (example below)
* Observe the output - make sure the query has been validated

### Example

I have been using a docker-compose project as a way to test this. I have also been using something similar to the following to test the functionality.

```bash
#!/bin/bash
mkdir data;
make build;
sudo rm ./data/*;
./bin/mtk-linux-amd64 dump local --port=3370 --user=root --password=root --host=localhost --data-export --data-path="s3://my-test-bucket" > data/_00.sql;
sudo chmod 777 data/*;
```

The output into `./data/_00.sql` of interest is as follows:

```SQL
LOAD DATA FROM S3 FILE 'S3-ap-southeast-2://my-test-bucket/block_content.csv' INTO TABLE `block_content` FIELDS TERMINATED BY ',' ENCLOSED BY '"' LINES TERMINATED BY '\n';
```

A sample of the query being executed, which not indicated but does support `WHERE` commands on generation.

```
SELECT `id`, `revision_id`, `type`, `uuid`, `langcode` FROM `block_content` INTO OUTFILE S3 's3://my-test-bucket/block_content.csv' FIELDS TERMINATED BY ',' ENCLOSED BY '"' LINES TERMINATED BY '\n' MANIFEST ON OVERWRITE ON;
```

## What is the rollout plan?

In spite of mysql instances and permissions that may need to be configured, the functionality is completely opt-in, and the following steps should be executed.

* Merge the pull request - the logic for queries has not been automatically altered as new functionality is strictly opt-in.
* Configure the mysql images and/or database sources as required. Work TBD.
* Add the new flags into your workflow, and test your standard workflow.

## Links

* https://docs.aws.amazon.com/prescriptive-guidance/latest/archiving-mysql-data/export-data-by-using-select-into-outfile-s3.html
* https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/AuroraMySQL.Integrating.LoadFromS3.html
